File: python/quick_sort.py
def quick_sort(arr, low, high, step=[1]):
    if low < high:
        pi = partition(arr, low, high)
        print(f"Iterasi {step[0]}: {arr}")
        step[0] += 1

        quick_sort(arr, low, pi-1)
        quick_sort(arr, pi+1, high)

def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i+1], arr[high] = arr[high], arr[i+1]
    return i+1

n = int(input("Jumlah data: "))
data = list(map(int, input().split()))

quick_sort(data, 0, len(data)-1)
print("Hasil:", data)

File: cpp/quick_sort.cpp
#include <iostream>
using namespace std;

void print(int arr[], int n) {
    for(int i=0;i<n;i++) cout << arr[i] << " ";
    cout << endl;
}

int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = low - 1;

    for(int j=low;j<high;j++) {
        if(arr[j] < pivot) {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i+1], arr[high]);
    return i+1;
}

void quickSort(int arr[], int low, int high, int n, int &step) {
    if(low < high) {
        int pi = partition(arr, low, high);

        cout << "Iterasi " << step++ << ": ";
        print(arr, n);

        quickSort(arr, low, pi-1, n, step);
        quickSort(arr, pi+1, high, n, step);
    }
}

int main() {
    int n;
    cout << "Jumlah data: ";
    cin >> n;

    int arr[n];
    cout << "Masukkan data: ";
    for(int i=0;i<n;i++) cin >> arr[i];

    int step = 1;
    quickSort(arr, 0, n-1, n, step);

    cout << "Hasil sorting: ";
    print(arr, n);
}
